pkg/controller/backup_queue_controller.go

```diff
@@ -0,0 +1,375 @@
+		if foundEarlierRunnable {
+			log.Infof("Earlier queued backup %v is runnable, leaving %v queued", earlierRunnable, backup.Name)
+			return ctrl.Result{}, nil
+		}
+		log.Infof("Dequeueing backup %v, moving to ReadyToStart", backup.Name)
+		original := backup.DeepCopy()
+		backup.Status.Phase = velerov1api.BackupPhaseReadyToStart
+		backup.Status.QueuePosition = 0
+		if err := kube.PatchResource(original, backup, r.Client); err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "error updating Backup status to %s", backup.Status.Phase)
+		}
+		log.Debug("Updating queuePosition for remaining queued backups")
+		queuedBackups := lister.orderedQueued()
+		newQueuePos := 1
+		for _, queuedBackup := range queuedBackups {
+			if queuedBackup.Name != backup.Name {
+				original := queuedBackup.DeepCopy()
+				queuedBackup.Status.QueuePosition = newQueuePos
+				if err := kube.PatchResource(original, &queuedBackup, r.Client); err != nil {
+					log.WithError(errors.Wrapf(err, "error updating Backup %s queuePosition to %v", queuedBackup.Name, newQueuePos))
+					return ctrl.Result{}, nil
+				}
+				newQueuePos++
+			}
+		}
+		return ctrl.Result{}, nil
+	default:
+		log.Debug("Backup is not New or Queued, skipping")
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// queuedBackupsLister manages a list of all backups Queued, ReadyToStart, or InProgress
+// with methods to return specific subsets as needed
+type queuedBackupsLister struct {
+	backups *velerov1api.BackupList
+}
+
+func (r *backupQueueReconciler) newQueuedBackupsLister(ctx context.Context, ns string) (*queuedBackupsLister, error) {
+	backupList := &velerov1api.BackupList{}
+	backups := []velerov1api.Backup{}
+	if err := r.Client.List(ctx, backupList, &client.ListOptions{Namespace: ns}); err != nil {
+		r.logger.WithError(err).Error("error listing backups")
+		return nil, err
+	}
+	for _, backup := range backupList.Items {
+		if backup.Status.Phase == velerov1api.BackupPhaseQueued ||
+			backup.Status.Phase == velerov1api.BackupPhaseInProgress ||
+			backup.Status.Phase == velerov1api.BackupPhaseReadyToStart {
+			backups = append(backups, backup)
+		}
+	}
+	backupList.Items = backups
+	return &queuedBackupsLister{backupList}, nil
+}
+
+func (l *queuedBackupsLister) earlierThan(queuePos int) ([]velerov1api.Backup, int) {
+	backups := []velerov1api.Backup{}
+	runningCount := 0
+	for _, backup := range l.backups.Items {
+		// InProgress and ReadyToStart backups have QueuePosition==0
+		if backup.Status.QueuePosition < queuePos {
+			backups = append(backups, backup)
+		}
+		// InProgress and ReadyToStart backups count towards the concurrentBackups limit
+		if backup.Status.Phase == velerov1api.BackupPhaseInProgress ||
+			backup.Status.Phase == velerov1api.BackupPhaseReadyToStart {
+			runningCount++
+		}
+	}
+	return backups, runningCount
+}
+
+func (l *queuedBackupsLister) queued() []velerov1api.Backup {
```

Comment:
Why do we have both `queued()` and `orderedQueued()`? I think `queued()` could be replaced by `orderedQueued()` in all the usages.